docs/datastorage/nbt.md

```diff
@@ -0,0 +1,98 @@
+  - However, lists of lists can mix and match different list types. So a list of two lists, where the first one is a list of strings and the second one is a list of bytes, is allowed.
+- There are special **array** types that are different from lists, but follow their scheme of containing elements in square brackets. There are three array types:
+  - Byte arrays, denoted by a `B;` at the beginning of the array. Example: `[B;0b,30b]`
+  - Integer arrays, denoted by a `I;` at the beginning of the array. Example: `[I;0,-300]`
+  - Long arrays, denoted by an `L;` at the beginning of the array. Example: `[L;0l,240l]`
+- Trailing commas in lists, arrays and compound tags are allowed.
+
+## NBT Files
+
+Minecraft uses `.nbt` files extensively, for example for structure files in [datapacks][datapack]. Region files (`.mca`) that contain the contents of a region (i.e. a collection of chunks), as well as the various `.dat` files used in different places by the game, are NBT files as well.
+
+NBT files are typically compressed with GZip. As such, they are binary files and cannot be edited directly.
+
+## NBT in Code
+
+Like in JSON, all NBT objects are children of an enclosing object. So let's create one:
+
+```java
+CompoundTag tag = new CompoundTag();
+```
+
+We can now put our data into that tag:
+
+```java
+tag.putInt("Color", 0xffffff);
+tag.putString("Level", "minecraft:overworld");
+tag.putDouble("IAmRunningOutOfIdeasForNamesHere", 1d);
+```
+
+Several helpers exist here, for example, `putIntArray` also has a convenience method that takes a `List<Integer>` in addition to the standard variant that takes an `int[]`.
+
+Of course, we can also get values from that tag:
+
+```java
+int color = tag.getInt("Color");
+String level = tag.getString("Level");
+double d = tag.getDouble("IAmRunningOutOfIdeasForNamesHere");
+```
+
+Number types will return 0 if absent. Strings will return `""` if absent. More complex types (lists, arrays, compounds) will throw an exception if absent.
+
+As such, we want to safeguard by checking if a tag element exists:
+
+```java
+boolean hasColor = tag.contains("Color");
+boolean hasColorMoreExplicitly = tag.contains("Color", Tag.TAG_INT);
+```
+
+The `TAG_INT` constant is defined in `Tag`, which is the super interface for all tag types. Most tag types besides `CompoundTag` are mostly internal, for example `ByteTag` or `StringTag`, though the direct `CompoundTag#get` and `#put` methods can work with them if you ever stumble across some.
+
+There is one obvious exception, though: `ListTag`s. Working with these is special because when getting a list tag through `CompoundTag#getList`, you must also specify the list type. So getting a list of strings, for example, would work like this:
+
+```java
+ListTag list = tag.getList("SomeListHere", Tag.TAG_STRING);
+```
+
+Similarly, when creating a `ListTag`, you must also specify the list type during creation:
+
+```java
+ListTag list = new ListTag(List.of(), Tag.TAG_STRING);
```

Comment:
Weird example due to how `ListTag` works. The type is mutable, so an empty list would have a tag type of `0` rather than the type of the object.